Pre-8.0.0 adjustments

jansiegel opened the pull request:
### Context
- Bump the Handsontable version to `8.0.0-beta.2`
- Make the `Handsontable.DefaultSettings` adjustment needed for the migration to `8.0.0`
- Fix the `npm audit` errors
- Update `@vue/test-utils` to a stable release and replace a deprecated option with the recommended one

### How has this been tested?
Tested manually and by running the test cases.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

jansiegel committed:
- Bump the Handsontable version to 8.0.0-beta.2
- Make the 'Handsontable.DefaultSettings' adjustment needed for the migration to 8.0.0
- Fix the npm audit errors
- Update '@vue/test-utils' to a stable release and replace a deprecated option with the recommended one

aninde commented:
The demo doesn't work https://github.com/handsontable/vue-handsontable-official/pull/180#issuecomment-660971801. I can't test.
<img width="371" alt="Screenshot 2020-07-21 at 11 08 30" src="https://user-images.githubusercontent.com/30186818/88035511-9073a700-cb42-11ea-8428-c187acd34aef.png">

jansiegel commented:
@aninde 
> The demo doesn't work [#180 (comment)](https://github.com/handsontable/vue-handsontable-official/pull/180#issuecomment-660971801). I can't test.

It looks like an issue with codesandbox, so you can try to test it locally.

jansiegel committed:
- Replace Handsontable.DefaultSettings type to Handsontable.GridSettings

jansiegel commented:
https://github.com/handsontable/vue-handsontable-official/pull/180/commits/1994375a35e23b8a3cebe7cf8c74b124a71baf51 contains changes for handsontable/handsontable#7145, so it needs to be tested with the current Handsontable `develop` (https://github.com/handsontable/handsontable/commit/aa00b372c757653e98a762bc8a7d76c2d3e19344 +).

aninde commented:
Demo built as @jansiegel requested https://jsfiddle.net/aninde/30u6t2fw/  Here Vanilla Handsontable from the same build. https://jsfiddle.net/aninde/vyks3x6h/ without this regression.

**Regression bug 1**
After `Insert operations` or`Remove columns` undo doesn't work, there is no such option at all in the context menu. 
Undo work after removing rows from the contextMenu, but not after removing colums.

![undo-ff](https://user-images.githubusercontent.com/30186818/88291383-7085e400-ccf8-11ea-95b3-5ccbb6344235.gif)

**Regression bug 2**
Wrong data are displaying after undoing removal of rows https://jsfiddle.net/aninde/b5fkyLaq/.
![Kapture 2020-07-23 at 16 09 01](https://user-images.githubusercontent.com/30186818/88296622-3d931e80-ccff-11ea-87f2-d705c45fc181.gif)

jansiegel commented:
> **Regression bug**

Not really a regression (as it happens in the current version as well: https://jsfiddle.net/js_ziggle/ejabsc37/1/), but a bug for sure. @aninde could you report a new issue for the problem?

aninde commented:
> @aninde could you report a new issue for the problem?

Of course. 

Regression 2 it also occurred before, but in a different form. In my opinion, with the v8 is worse. https://jsfiddle.net/aninde/wyz67Lnd/ v.7.4.2 for comparing to https://jsfiddle.net/aninde/b5fkyLaq/.
![stable-reg2](https://user-images.githubusercontent.com/30186818/88297198-e93c6e80-ccff-11ea-9492-afa4626e87cc.gif)

aninde commented:
**Regression 3**
Indexes are lost. While I'm Inserting on moved row, insert works on the wrong row. It's regression comparing to stable Vanilla and 7.4.2 with Vue-Handsontable.
The same bug is present after inserting col on moved column https://jsfiddle.net/aninde/u6y1c59z/

v8 https://jsfiddle.net/aninde/9azqx2jw/
![v8](https://user-images.githubusercontent.com/30186818/88302767-869aa100-cd06-11ea-8f3c-bfb7b4699726.gif)
v7 https://jsfiddle.net/aninde/45Lps81t/
![v7](https://user-images.githubusercontent.com/30186818/88302772-87cbce00-cd06-11ea-86d1-b4bb3a6b1761.gif)

v8, insert on moved column
![moved-col](https://user-images.githubusercontent.com/30186818/88386106-003a9980-cdb0-11ea-81f2-191e71f3e79c.gif)

aninde commented:
**Regression 4**
Doubled border on bottom edge of colHeader https://jsfiddle.net/aninde/h29nocb0/ - this, currently tested version.
https://jsfiddle.net/aninde/0khw4dgj/ - `v.7.4.2` - fixed, 
vanilla from the same build of Handsontable https://jsfiddle.net/aninde/cob8zv96/ `develop` fixed.

jansiegel commented:
@aninde 
> **Regression 4**
> Doubled border on bottom edge of colHeader https://jsfiddle.net/aninde/h29nocb0/ - this, currently tested version.

> vanilla from the same build of Handsontable https://jsfiddle.net/aninde/cob8zv96/ `develop` fixed.

I can see the same wider border in both of these demos, which would indicate it's a not a wrapper issue.

jansiegel committed:
- Make the wrapper NOT call updateSettings with the table data updates, instead update the index mappers, as the data itself is synchronized by reference,
- Add a simple diff function to prevent updating with the props which were not changed
  #185

jansiegel committed:
- Change the 'countRows' and 'countCols' to their source counterparts. #185

jansiegel committed:
Modify the handsontable's indexMapper handling in case of a data update.

jansiegel committed:
Merge branch 'feature/issue-185' into feature/8.0.0-beta2-adjustments

jansiegel committed:
- Rename the Handsontable's index mapper handling function.

jansiegel commented:
@wojciechczerniak I merged the logic from #185 here, because it turned out to need `handsontable@8.0.0-beta.2+` to work properly anyway.

wojciechczerniak commented on src/HotTable.vue:
```diff
@@ -86,6 +98,38 @@
+        const indexMapperRowCount = this.hotInstance.rowIndexMapper.getNumberOfIndexes();
+        const indexMapperColumnCount = this.hotInstance.columnIndexMapper.getNumberOfIndexes();
+
+        if (data && data.length !== indexMapperRowCount) {
+          if (data.length < indexMapperRowCount) {
+            for (let r = data.length; r < indexMapperRowCount; r++) {
+              rowsToRemove.push(r);
+            }
+
+            this.hotInstance.rowIndexMapper.removeIndexes(rowsToRemove);
+
+          } else {
+            this.hotInstance.rowIndexMapper.insertIndexes(indexMapperRowCount - 1, data.length - indexMapperRowCount);
+          }
+        }
+
+        if (data && data[0] && data[0]?.length !== indexMapperColumnCount) {
+          if (data[0].length < indexMapperColumnCount) {
+            for (let c = data[0].length; c < indexMapperColumnCount; c++) {
+              columnsToRemove.push(c);
+            }
+
+            this.hotInstance.columnIndexMapper.removeIndexes(columnsToRemove);
+
+          } else {
+            this.hotInstance.rowIndexMapper.insertIndexes(indexMapperColumnCount - 1, data[0].length - indexMapperColumnCount);
+          }
+        }
```
Gather all changes to index mappers and run them in batch. Should result in a single render

```suggestion
        if (data && data.length !== indexMapperRowCount) {
          if (data.length < indexMapperRowCount) {
            for (let r = data.length; r < indexMapperRowCount; r++) {
              rowsToRemove.push(r);
            }
          }
        }
        
        this.hotInstance.batch(() => {
            if (rowsToRemove.length > 0) {
              this.hotInstance.rowIndexMapper.removeIndexes(rowsToRemove);
            } else {
              this.hotInstance.rowIndexMapper.insertIndexes(indexMapperRowCount - 1, data.length - indexMapperRowCount);
            }
            
            if (columnsToRemove.length > 0) {
              this.hotInstance.columnIndexMapper.removeIndexes(columnsToRemove);
            } else {
              this.hotInstance.rowIndexMapper.insertIndexes(indexMapperColumnCount - 1, data[0].length - indexMapperColumnCount;
            }
        });
```

jansiegel replied:
> Should result in a single render

@wojciechczerniak I can run them in a batch, but I don't think they trigger a render anyway? https://jsfiddle.net/js_ziggle/ndsg1906/

wojciechczerniak replied:
Sorry, my bad. I was thinking about `updateCache` https://jsfiddle.net/mfp316rq/  But here we always have just one call for each indexMapper so I guess it doesn't matter.

wojciechczerniak commented on src/HotTable.vue:
```diff
@@ -86,6 +98,38 @@
+            this.hotInstance.columnIndexMapper.removeIndexes(columnsToRemove);
+
+          } else {
+            this.hotInstance.rowIndexMapper.insertIndexes(indexMapperColumnCount - 1, data[0].length - indexMapperColumnCount);
```
Copy paste mistake?
```suggestion
            this.hotInstance.columnIndexMapper.insertIndexes(indexMapperColumnCount - 1, data[0].length - indexMapperColumnCount);
```

wojciechczerniak commented on src/HotTable.vue:
```diff
@@ -86,6 +98,38 @@
+        const rowsToRemove: number[] = [];
+        const columnsToRemove: number[] = [];
+        const indexMapperRowCount = this.hotInstance.rowIndexMapper.getNumberOfIndexes();
+        const indexMapperColumnCount = this.hotInstance.columnIndexMapper.getNumberOfIndexes();
```
🤔 Are you sure this will work when `columns` are defined? Or Array of Objects is used or `schema` is defined?

There might be a slight misalignment between getNumberOfIndexes and sourceCount

jansiegel replied:
@wojciechczerniak You're right, this should be probably only done for arrays of arrays, modifications on arrays of objects should be done through `updateSettings`. It's impossible to add/remove columns on AoO with the UI though, so it would affect only changes done by reference from the outside.

jansiegel committed:
- Apply CR suggestions
- Correct the logic for AoO-based datasets
- Add additional tests
